spec/Spec/BlsEC.cry

```diff
@@ -0,0 +1,226 @@
+/**
+ * This module contains functions that perform ECC Point operations.
+ * These functions are specific to the BLS curve, and specifially G1.
+ *
+ * The Deneb Python specs use the py_ecc library for the same operations.
+ * This modules functions are based off of py_ecc.
+ * @see https://github.com/ethereum/py_ecc/blob/main/py_ecc/bls12_381/bls12_381_curve.py
+ */
+module Spec::BlsEC where
+
+import Common::Utils
+
+/**
+ * Size of a G1 point
+ */
+type G1Bits = 384
+
+/**
+ * Serialized G1 point
+ */
+type Bytes48 = [G1Bits]
+
+/**
+ * G1 elliptic curve point
+ */
+type G1Point =
+    { xCoord : Bytes48
+    , yCoord : Bytes48
+    }
+
+/**
+ * G1 generator
+ * @see https://github.com/ethereum/py_ecc/blob/main/py_ecc/bls12_381/bls12_381_curve.py#L35
+ */
+G1 : G1Point
+G1 =
+    { xCoord = 3685416753713387016781088315183077757961620795782546409894578378688607592378376318836054947676345821548104185464507
+    , yCoord = 1339506544944476473020471379941921221584933875938349620426543736416511423956333506472724655353366534992391756441569
+    }
+
+/**
+ * The BLS12-384 curve is defined as: y^^2 = x^^3 + 4 over the
+ *  finite field Fp, where p is the prime defined below.
+ * @see https://hackmd.io/@Wimet/S1dvK2NBh#The-Curve
+ */
+Fp : Bytes48
+Fp = 4002409555221667393417789825735904156556882819939007885332058136124031650490837864442687629129015664037894272559787
+
+/**
+ * Determines whether or not the given point is on the G1 curve.
+ * The point must satisfy the EC equation: y^^2 = x^^3 + 4
+ * ```repl
+ * :prove g1_is_valid_point G1
+ * :prove g1_is_valid_point G1_2P
+ * :prove g1_is_valid_point G1_3P
+ * :prove g1_is_valid_point G1_4P
+ * :prove g1_is_valid_point G1_5P
+ * :prove g1_is_valid_point G1_6P
+ * ```
+ */
+g1_is_valid_point : G1Point -> Bit
+g1_is_valid_point point =
+    (y^^2 - x^^3) % modulus == 4
+    where
+        x = toInteger point.xCoord
+        y = toInteger point.yCoord
+        modulus = toInteger Fp
+
+/**
+ * Double a G1 point
+ * @see https://github.com/ethereum/py_ecc/blob/main/py_ecc/bls12_381/bls12_381_curve.py#L84
+ */
+g1_double : G1Point -> G1Point
+g1_double point = newPoint where
+    x = toInteger point.xCoord
+    y = toInteger point.yCoord
+    modulus = toInteger Fp
+    m = div (3 * x^^2) (2 * y) modulus
```

Comment:
suggestion: consider renaming `div` to `div_mod` if you keep this function. `modulus` is pretty clear but it's not immediately obvious why div has 3 parameters.

Reply:
I like this suggestion. I'm thinking though if I change it to use `Z` then I don't need the extra parameter, so hopefully that is more clear. I only used the name `div` because it matches the Python spec, but I prefer your suggestion, even as I go with using `Z`.